File: 2381-shifting-letters-ii/2381-shifting-letters-ii.cpp
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        int n = s.size();
        vector<int>pref(n + 1);
        
        for (auto x: shifts) {
            pref[x[0]] += (x[2] == 1 ? 1 : -1);
            pref[x[1] + 1] += (x[2] == 1 ? -1 : 1);
        }
        
        for (int i = 1; i <= n; i++)
            pref[i] += pref[i - 1];
        // a%b + b)%b 
        for (int i = 0; i < n; i++)
            s[i] = char('a' + ((s[i] - 'a' + pref[i]) % 26 + 26) % 26);
        
        return s;
    }
};

File: 2381-shifting-letters-ii/2381-shifting-letters-ii.py
class Solution:
    def shiftingLetters(self, s: str, shifts: List[List[int]]) -> str:
        n = len(s)
        pref = [0] * (n + 1)
        
        for x in shifts:
            pref[x[0]] += (1 if x[2] == 1 else -1)
            pref[x[1] + 1] += (-1 if x[2] == 1 else 1)
            
        for i in range(1, n + 1):
            pref[i] += pref[i - 1];
        
        s = list(s)
        for i in range(n):
            s[i] = chr((ord(s[i]) - ord('a') + pref[i]) % 26 + ord('a'))
            
        return "".join(s)